onnxruntime/core/providers/cuda/cuda_graph.cc

```diff
@@ -29,40 +73,70 @@ void CUDAGraph::CaptureBegin() {
   CUDA_CALL_THROW(cudaStreamBeginCapture(stream_, cudaStreamCaptureModeGlobal));
 }
 
-void CUDAGraph::CaptureEnd() {
+void CUDAGraphManager::CaptureEnd() {
+  if (!IsGraphCaptureAllowedOnRun()) {
+    LOGS_DEFAULT(INFO) << "Skipping graph capture for cuda_graph_annotation_id " << cuda_graph_annotation_id_;
+    return;
+  }
+
   CUDA_CALL_THROW(cudaStreamEndCapture(stream_, &graph_));
   if (graph_ == NULL) {
     ORT_THROW("CUDAGraph::CaptureEnd: graph_ is NULL");
   }
-
   has_graph_ = true;
-  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
-  has_graph_exec_ = true;
+
+  cudaGraphExec_t graph_exec = NULL;
+  CUDA_CALL_THROW(cudaGraphInstantiate(&graph_exec, graph_, NULL, NULL, 0));
   CUDA_CALL_THROW(cudaGraphDestroy(graph_));
   has_graph_ = false;
+
+  cuda_graph_set_.Put(cuda_graph_annotation_id_, graph_exec);
+  has_graph_exec_ = true;
 }
 
-Status CUDAGraph::Replay() {
+Status CUDAGraphManager::Replay() {
+  if (!IsGraphCaptureAllowedOnRun()) {
```

Comment:
Shall not call this function when not allowed. May change to:
ORT_ENFORCE(!IsGraphCaptureAllowedOnRun());